### PR TITLE
DEV: generalized cross-validation smoothing spline

### PR DESCRIPTION
**Pull request by egorz734**
# Introduction
A new function `make_smoothing_spline` is added.

Most of the equations and references to the algorithms are described in [1] ([link](https://www.hse.ru/ba/am/students/diplomas/page3.html/620910604) only in Russian).

The basic idea that lies behind the algorithm is straightforward:
Suppose we have some noise data - data points (x, y) and we want to fit a smoothing spline $f$ in it. In order to do that, we solve the minimization problem for the given $\lambda \geq 0$:
$$\sum\limits_{i=1}^n w_i |y_i - f(x_i)|^2 + \lambda\int\limits_{x_1}^{x_n} (f^{(2)}(u))^2 d u \rightarrow \mathrm{min}$$
$\lambda$ here controls the tradeoff between the smoothness of the curve and its proximity to the data. To find an “optimal” $\lambda$ we solve another minimization problem:
$$V(\lambda)=\dfrac{1}{n} \dfrac{ ||(I - A)y||^2}{\left(\dfrac{1}{n} \mathrm{Tr}(I - A)\right)^2}\rightarrow \mathrm{min}$$
where $A$ is a smoothing matrix.

# Results
To find a smoothing spline for some set of data points with noise $(x,y)$ we call the function: `make_smoothing_spline(x,y)` which returns a `BSpline` corresponding to a smoothing spline.

`_make_smoothing_spline` is used to find a smoothing spline with give parameter $\lambda$.

# References
[1] E. Zemlyanoy, E. Burovski, "Generalized cross-validation smoothing splines", 2022.

### CODE REVIEW

**Comment by ev-br** [comment]
Two quick notes: 
- this is a clean-room reimplementation of the GCV spline smoothing algorithm. The canonical implementation, gcvspl.f, is packaged in e.g. Matlab/Octave, but is license-incompatible, see https://github.com/scipy/scipy/issues/6730#issuecomment-298513134 for a discussion.
- Ref [1] in the OP is Egor's BSc thesis, so he's the single author :-). 

While I'll do technical review of this PR, more reviews are most welcome. If you use GCV splines in Matlab or R's `smooth.spline`, please take a look and let us know what you think!

**Comment by dschmitz89** [comment]
Saw this coincidentally. Are these new splines what is also known as generalized additive models (GAM)? The math looks like it to me. For that, I had good experience with [pyGAM](https://github.com/dswah/pyGAM) in case you need/want to compare to another python implementation.

**Comment by ev-br** [comment]
@dschmitz89 Looks related indeed. I'd expect this to be lower level, and certainly much more narrow in functionality. 
Will need to check the references in https://pygam.readthedocs.io/en/latest/notebooks/tour_of_pygam.html#Generalized-Additive-Models,-in-general;  will take a look starting from Refs 4 and 7. I've no experience with GANs, and I'm not a statistician at all, so help from that angle would be very much appreciated!

**Comment by rkern** [comment]
Splines like this are frequently-used models in GAMs (as the individual nonlinear Models that are summed Additively), and techniques like CV are common to select the tradeoff parameters. The "Generalized" part of GAM refers to having different link functions (or the "data-fidelity term of the loss function" in ML terms) whereas what's implemented here just uses the usual squared error link function/loss. So yes, there's a relationship to the "M" (and the overall CV scheme), but not the "G" or "A" part of "GAM".

File: scipy/interpolate/_bsplines.py [hunk 1]
```diff
@@ -1521,3 +1523,532 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+    algorithm is based on [2].
+
+    Parameters
+    ----------
+    x : array_like, shape (n,)
+        Abscissas.
+    y : array_like, shape (n,)
+        Ordinates.
+    w : array_like, shape (n,), optional
+        Vector of weights.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1999]: Default is `np.ones_like(x)`?

File: scipy/interpolate/_bsplines.py [hunk 2]
```diff
@@ -1521,3 +1523,532 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+        
+        Notes
+        -----
+        The algorithm is based on the cholesky decomposition and, therefore,
+        in case matrix ``A`` is close to not positive defined, the function
+        raises LinalgError.
+
+        Both matrices ``A`` and ``B`` are stored in the way that ``X[-1]`` is
+        the main diagonal of the matrices, ``X[-2]`` is the second diagonal,
+        etc.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1800]: Is this LAPACK banded storage or something else?

File: scipy/interpolate/_bsplines.py [hunk 3]
```diff
@@ -1521,3 +1523,532 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+        Returns
+        -------
+        res : array, shape (4, n)
+            The result of the product :math:`X^T Y` stored in the banded way.
+
+        Notes
+        -----
+        As far as the matrices ``X`` and ``Y`` are 5-banded, their product 
+        :math:`X^T W Y` is 7-banded. It is also symmetric, so we can store only
+        unique bands.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1757]: typo: unique diagonals?

**Review by ev-br — CHANGES_REQUESTED** [review]
Initial cursory review. Overall looks good!
Did not read through the algebra yet, will do in a second pass.

**Comment by josef-pkt** [comment]
A question (instead of trying to understand the code)

How are the knots selected? Are there knots at every observation point?

The statsmodels GAM version assumes that we don't have too many knots (basis functions), otherwise the computation become memory intensive and fragile.

**Comment by egorz734** [comment]
> A question (instead of trying to understand the code)
> 
> 
> 
> How are the knots selected? Are there knots at every observation point?
> 
> 
> 
> The statsmodels GAM version assumes that we don't have too many knots (basis functions), otherwise the computation become memory intensive and fragile.
> 
> 

In a returned `BSpline` object the knots are selected like in a natural smoothing spline: $[x_1, x_1, x_1, x_1, x_2, \dots, x_{n-1}, x_n, x_n, x_n, x_n]$, so, yes they are in every observation point. 

About computation and memory: they both require O(n) (except for the minimization algorithm which is iterative).

**Comment by josef-pkt** [comment]
good, that's useful. 
This case cannot be handled well by using basis functions in a design matrix for regression as in standard spline based GAM.

**Comment by ev-br** [comment]
Why it cannot be handled, Josef? 
And how does statsmodels select the knot vector?

**Comment by josef-pkt** [comment]
> Why it cannot be handled, Josef?

Each basis function adds a column to the design matrix. So we need approximately nobs basis functions (IIRC) if we have around nobs knots. The design matrix will then have too many columns to work efficiently with standard (penalized) regression methods.
(Working with sparse matrices might be possible but not all supported splines are local and sparse.)

> And how does statsmodels select the knot vector?

two options:
- knot at every kth observation  (patsy uses this for the splines)
- knots equal spaced in x space  (that's the standard for "penalization splines" which statsmodels doesn't fully implement yet.)

statsmodels also has options for setting boundary knots, including knots that are possibly outside the x-space, so we can extrapolate in some range.

**Comment by ev-br** [comment]
> Each basis function adds a column to the design matrix. So we need approximately nobs basis functions (IIRC) if we have around nobs knots. The design matrix will then have too many columns to work efficiently with standard (penalized) regression methods.

What's the bottleneck: memory, CPU time or numerical errors?

> (Working with sparse matrices might be possible but not all supported splines are local and sparse.)

Ah, so for b-splines specifically in scipy >= 1.8.1 you'd be able to rely on BSpline.design_matrix which constructs it as a csr_array :-).

* knots equal spaced in x space (that's the standard for "penalization splines" which statsmodels doesn't fully implement yet.)

Would be able to provide some references? More knot selection methods is certainly on the todo list; certainly not in this PR though.

**Comment by josef-pkt** [comment]
for P-splines: Eilers and Marx, (main feature is difference penalty for neighboring coefficients instead of full penalization matrix for second derivatives.)

https://github.com/statsmodels/statsmodels/issues/4908#issuecomment-413995617
https://stats.stackexchange.com/questions/410515/how-different-are-restricted-cubic-splines-and-penalized-splines
main reference:
Eilers, Paul H. C.; Marx, Brian D. Flexible smoothing with B -splines and penalties. Statist. Sci. 11 (1996), no. 2, 89--121. doi:10.1214/ss/1038425655. https://projecteuclid.org/euclid.ss/1038425655

but, AFAIR, they have other articles with same or similar method.

> What's the bottleneck: memory, CPU time or numerical errors?

all of the above: matrices are around nobs x nobs if there is only one spline component, and much larger if there are several.
If the number of observations, nobs, is large, then having many (nobs, nobs) arrays requires large memory, inverting those is cpu intensive and possibly high collinearity makes optimization difficult and fragile.
With smaller sample sizes, the optimization problems where the main problem, based on what I remember from my experiments with it.

**Comment by lorentzenchr** [comment]
It might be of interest that scikit-learn has a [`SplineTransformer`](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.SplineTransformer.html#sklearn.preprocessing.SplineTransformer) which uses `BSpline` from scipy and has several knot selection methods, e.g. manual, uniform or quantile based.
Using every data point as knot gives a smoothing splines (spline interpolation). I think one of the main advantages of penalized splines is parsimonious and yet very flexible modelling, i.e. using (still many but) much less knots than data points.

One of the best references/books about splines (next to the seminal paper of Eilers & Marx) is certainly: Simon N. Woods "Generalized Additive Models: an introduction with R"

**Comment by SeanChen1992** [comment]
您的邮件我已收到，我会加班加点的看完您珍贵的来信，尽快给您回复的！谢谢啦！陈鑫

**Comment by ev-br** [comment]
Thanks for the references! It's now clear that adding a `t=...` kwarg should be one of the first things to consider for a follow-up (where the current behavior would become `t=None` or some such). For this PR specifically, ISTM it's best to keep it small(-ish) and focused on adding the very core GCV functionality.

**Comment by egorz734** [comment]
A quick overview of what is done:
1. Now the new interface is `make_smoothing_spline(x, y, w=None, lam=None)`, where `lam` is the regularization parameter.
2. Removed nesting and added more comments to make the code more understandable and easy-to-follow.
3. Fixed the references and docstrings (especially in the API docs).
4. Added an explanation of the Octave script that uses gcvspl.
5. Added a few more tests and checks for the input vectors shapes.

File: scipy/interpolate/_bsplines.py [hunk 4]
```diff
@@ -1521,3 +1523,532 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+def make_smoothing_spline(x, y, w=None):
+    r'''
+    Returns a smoothing cubic spline function using the GCV criteria.
+    A smoothing spline is found as a solution to the following minimization
+    problem:
+    
+    .. math::
+
+        \sum\limits_{i=1}^n w_i\abs{y_i - f(x_i)}^2 + 
+        \lambda\int\limits_{x_1}^{x_n} (f^{(2)}(u))^2 \diff u
```
[the remark this answers is not in the document]
Comment (egorz734) [reply scipy/interpolate/_bsplines.py:1798]: The weights are just `w`, not squared for the consistency with gcvspl.
Also left in the `Notes` the comment about `splrep`: lines 1834 - 1836.

File: scipy/interpolate/tests/test_bsplines.py [hunk 5]
```diff
@@ -1459,3 +1461,102 @@ def test_checkfinite(self):
...
+
+
+class TestSmoothingSpline:
+    #
+    # test make_smoothing_spline
+    #
+    def test_invalid_input(self):
+        np.random.seed(1234)
+        n = 100
+        x = np.sort(np.random.random_sample(100) * 4 - 2)
```
[the remark this answers is not in the document]
Comment (egorz734) [reply scipy/interpolate/tests/test_bsplines.py:1477]: Yes, missed that comment. Will fix in the next commit.

File: scipy/interpolate/_bsplines.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lorentzenchr) [reply scipy/interpolate/_bsplines.py:1798]: *Not* squaring weights seems sound to me, too.

**Comment by ev-br** [comment]
Could you rebase on main or merge main into your branch? Some CI failures are probably fixed in main.

**Comment by ev-br** [comment]
Mailing list discussion: https://mail.python.org/archives/list/scipy-dev@python.org/thread/Z6ZN7VLD7GZV3XXVE3J52LF34MVKROBY/

File: scipy/interpolate/_bsplines.py [hunk 6]
```diff
@@ -1561,3 +1563,432 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+    -------
+    lam : float
+        An optimal from the GCV criteria point of view regularization
+        parameter.
+
+    References
+    ----------
+    .. [1] G. Wahba, “Estimating the smoothing patameter” in Spline models
+        for observational data, Philadelphia, Pennsylvania: Society for
+        Industrial and Applied Mathematics, 1990, pp. 45–65.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1598]: Could you add a DOI link https://epubs.siam.org/doi/book/10.1137/1.9781611970128

File: scipy/interpolate/_bsplines.py [hunk 7]
```diff
@@ -1561,3 +1563,432 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+        References
+        ----------
+        .. [1] G. Wahba, “Estimating the smoothing patameter” in Spline models
+            for observational data, Philadelphia, Pennsylvania: Society for
+            Industrial and Applied Mathematics, 1990, pp. 45–65.
+        .. [2] M. F. Hutchinson and F. R. de Hoog, “Smoothing noisy data with spline 
+            functions,” Numerische Mathematik, vol. 47, no. 1, pp. 99–106, 1985.
+            :doi:`10.1007/BF01389878`
+        .. [3] E. Zemlyanoy, "Generalized cross-validation smoothing splines",
+            2022.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1752]: A link to the HSE collection would be useful, otherwise it's not searchable.

File: scipy/interpolate/_bsplines.py [hunk 8]
```diff
@@ -1561,3 +1563,432 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+        for k in range(n):
+            if k != i:
+                pp *= (x[i] - x[k])
+        res[i] = 1. / pp
+    return res
+
+
+def make_smoothing_spline(x, y, w=None, lam=None):
+    r'''
+    Returns a smoothing cubic spline function using the GCV criteria [1].
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1830]: Strictly speaking, this line is stale since you now accept `lam` as user input.

File: scipy/interpolate/_bsplines.py [hunk 9]
```diff
@@ -1561,3 +1563,432 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+    lam : float, (:math:`\lambda \geq 0`), optional
+        Regularization parameter. If ``lam`` is None, then it is found from
+        the GCV criteria. Default is None.
+    
+    Returns
+    -------
+    func : a BSpline object.
+        A callable representing a spline in the B-spline basis 
+        as a solution of the problem of smoothing splines using
+        the GCV criteria [1].
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1864]: Also here: it now returns a spline given the smoothing parameter `lam`.

File: scipy/interpolate/_bsplines.py [hunk 10]
```diff
@@ -1561,3 +1563,432 @@ def make_lsq_spline(x, y, t, k=3, w=None, axis=0, check_finite=True):
...
+    
+    Notes
+    -----
+    The original implementation was introduced by Woltring in FORTRAN [2], but
+    because of the license issues cannot be used in the SciPy source code.
```
Comment (ev-br) [inline scipy/interpolate/_bsplines.py:1869]: Consider rephrasing: this is  a clean-room reimplementation of the code in [2], correct?

**Review by ev-br — COMMENTED** [review]
Needs a rebase on main.

**Comment by rgommers** [comment]
Looks like nice work @egorz734!

I have an API design question (thanks for pointing me here @ev-br). I realize that `make_smoothing_spline` follows a pattern set by `make_interp_spline`, so my question is probably more about the latter. On the mailing list @ev-br posted this table of recommended APIs: https://gist.github.com/ev-br/fcd25570c3f332faefd521e6d3c87510. It looks to me like it's mostly class-based; there are a couple of flavors of API style in `scipy.interpolate`, but `make_interp_spline` is fairly unique and it seems to me that it has made the overall `interpolate` API less coherent. Was there a discussion on this (if so, I missed it or forgot by now)? 

If `make_interp_spline` is a factory function for `BSpline` instances, I'm not sure why it could not have been the regular `BSpline` constructor. There are already a number of constructors on `BSpline` itself:
- `__init__`
- `construct_fast`
- `from_power_basis`

Wouldn't it have made more sense for `make_interp_spline` to be what `BSpline.__init__` does and then move the `(t, c, k)` constructor to a method? That would move us towards a more uniform API for `interpolate`, rather than further away from it.

_Disclaimer: it is very well possible that I missed a fundamental reason for why something like that is not possible, I don't use these APIs a lot._

**Comment by ev-br** [comment]
I now realize that I never replied to @rgommers 's comment in that "let's deprecate interp2d" mailing list thread. Will reply here and link from the email thread then. 

The canonic version of the table is now at https://scipy.github.io/devdocs/tutorial/interpolate.html (will get even more official in the 1.10 release :-)).

TL;DR I really don't like the idea of having make_interp_spline and its relatives alternative constructors for the BSpline class.
(am on record of being opinionated, yes).

The fundamental reason is the separation of concerns. BSpline class contains knots and coefficients and knows nothing about the data. The factory function takes the data and constructs the corresponding BSpline instance. 

In this sense, BSpline is something a user does not need to care about, really. it's a duck-typed object which can has the `__call__` method and a couple of other useful methods like `.derivative`. It is similar to `PPoly`, only represents polynomials in a different basis.

The alternative constructors (`BSpline.construct_fast`, `BSpline.from_power_basis`) also deal with the low-level BSpline interface, known nothing about the data and have nothing to do with data. 

Adding alternative constructors which take in the data is in principle possible, but 
- it will break the separation between representation (current BSpline), and intepolation (current `make_interp_spline`)
- then we'll have a growing collection of them. At the moment it's `make_interp_spline` and `make_lsq_spline`, this PR would add `make_smooth_spline`, the future planned work (FITPACK alternatives) would either add more options to the latter or more alternative constructors. 

Indeed, this was not discussed much. There was some discussions, like  _Not sure I like the make_* names... Don't have better suggestions ATM though._ in https://github.com/scipy/scipy/pull/3174#issuecomment-120959918
but then nobody had better suggestions and lazy consensus won :-). And then the names got pinned by being released.

As to the inconsistency with `CubicSpline` et al, the reason is mostly historical :-). The fact that e.g. `CubicSpline` is a `PPoly` subclass is, uhm, an artifact of history where a reviewer did not like the design but not enough to block an otherwise good PR :-).
It seems to me this pattern of submitter-reviewer interaction has repeated several times over the history of this subpackage :-).
I still prefer factory functions to changing the constructor signature at will (LSP and all), but then I mean, I'm ready to code around almost any not-completely-insane API.

**Comment by rgommers** [comment]
> The fundamental reason is the separation of concerns. BSpline class contains knots and coefficients and knows nothing about the data. 

That's only technically true - since `__call__` does that `x` it does know about the data axis, it's not only a "knots and coefficients" object. 

> it will break the separation between representation (current BSpline), and intepolation (current `make_interp_spline`)

I'd say that (a) it's already broken because of `__call__`, and (b) to me, "interpolation" is the combined process including getting the final interpolated output.



> As to the inconsistency with `CubicSpline` et al, the reason is mostly historical :-). The fact that e.g. `CubicSpline` is a `PPoly` subclass is, uhm, an artifact of history where a reviewer did not like the design but not enough to block an otherwise good PR :-).
> It seems to me this pattern of submitter-reviewer interaction has repeated several times over the history of this subpackage :-).

I'm not inclined to add to that here, but I would urge more work on the overall picture. A "factory function plus class that exposes low-level representation details" is not a good high-level API for the average user. It's a low-level/advanced API. Things like `griddata` _are_ good API for the average user. So I'd really like to see an overall approach of how we'd like to converge to a recommended high-level API (as in, the one in your table). I don't have a preference for whether that is a class-based or functional one, but I do have a preference for it being mostly consistent and understandable.

**Comment by lorentzenchr** [comment]
> So I'd really like to see an overall approach of how we'd like to converge to a recommended high-level API

The `interpolate` module reminds me a lot of scikit-learn. An interpolator is like an estimator, i.e. a class that is instantiated with certain parameters (number of knots, polynomial degree, penalty/smoothing parameter, ..), a `fit` method using the data (x, y) and a `predict` method taking (potentially new) data x (without response y).

I personally (and by my role, of course) like the semantical simplicity of this API design. But I don‘t know whether it is a good fit for scipy.interpolate given its current API(s).

**Comment by ev-br** [comment]
ISTM there are two separate topics here: one is specifically the GCV spline functionality and the other one is a bigger topic of an overall API consistency.  
The former can be finished up quickly, even for the imminent 1.10 release.
The latter one seems to be calling for a larger (re)design/planning, with backwards compatibility concerns to work out. 
It definitely cannot be done is the 1.10 timeframe. 
So, I'd suggest we keep these two separate.  

Specifically, how about we
- merge this PR in its current state for 1.10 (after @egorz734 fixes the last technical issues), and
- open a separate issue to discuss the grand plan for the scipy.interpolate API.

An alternative is of course to postpone this PR to 1.11 or later.

For the record, I don't agree with all what Ralf said above, and the connection to the scikit-learn API is intriguing indeed. However let's separate all this from the specific bits of functionality this PR brings in.
WDYT?

**Comment by rgommers** [comment]
Nice connection to scikit-learn.

Merging this PR for 1.10 seems good to me. Given the recent mailing list discussion and this PR adding new API surface, it triggered the conversation here - but fundamentally this doesn't change it much, so let's not block good work on a long overdue big picture design.

**Comment by ev-br** [comment]
OK, musl CI failure is unrelated, all other failures are resolved, the tolerance for the  gcvspl comparison is explained to be due to different minimization primitives: for gcvspl it's a black box, for scipy it's our standard minimize_scalar; the end results do not and are not expected to agree to anything close to machine precision.

General reception of this as new feature is positive, and a more general discussion of scipy.interpolate API redesign is separate. 

So merging, thank you @egorz734 for implementing it, thanks all for reviews and discussion.
